### ISSUE TITLE
updates: Work around the "whilst offline" in Ubuntu & Debian

### ISSUE THREAD

**Issue by garrett** [cockpit-project/cockpit#16963]
The Updates page has an "emergent bug" between PackageKit, NetworkManager, and systemd-networkd.

As seen on
- https://github.com/cockpit-project/cockpit/issues/8477
- https://github.com/cockpit-project/cockpit/issues/16333#issuecomment-1004070722
- and elsewhere (including other bugs)

As our users will hit this bug on Ubutntu and Debian systems, we should provide a simple workaround.

This is not a bug in Cockpit, but this is a UX bug that affects Cockpit.

The solution is to create a fake network to make all of the subsystems happy:
```
nmcli con add type dummy con-name fake ifname fake0 ip4 1.2.3.4/24 gw4 1.2.3.1
```

We should add the workaround as a simple solution with a short description of the problem and the solution on the Software Updates page when this problem occurs.

If upstream fixes their systems to not trigger this issue, then it'll be transparently fixed, even with our workaround. Until then, a workaround like this will save a lot of people headaches.

**Comment by garrett** [comment]
We could have the workaround in Cockpit's Software Updates page where we augment the error to include the workaround, like this:

> [Icon]
> ### Network issues fetching updates
> 
> PackageKit believes your computer is offline. Either reload the page or apply a workaround(?) to make PackageKit communicate with NetworkManager. 
> 
> [[Reload]] [Apply workaround]



The (?) would be a popover that says the following:


> The workaround will add a fake network to your system using the following command:
> ```
> nmcli con add type dummy con-name fake ifname fake0 ip4 1.2.3.4/24 gw4 1.2.3.1
> ```


Reload would be the default option; applying a workaround would be a secondary action.

**Comment by barry-luijten** [comment]
Hi, I ran into this issue on Ubuntu 20.04, but creating the fake connection did not resolve the issue.
In order to expose VMs on my network, I have created a bridge device. Might this cause the issue?
```
# nmcli device
DEVICE  TYPE      STATE      CONNECTION
br0     bridge    unmanaged  --
eno1    ethernet  unmanaged  --
lo      loopback  unmanaged  --

# nmcli conn show
NAME  UUID                                  TYPE   DEVICE
fake  36b1c3c4-33df-47d4-bece-cc5366759997  dummy  --

```

**Comment by Bazimek** [comment]
Hello, will this : nmcli con add type dummy con-name fake ifname fake0 ip4 1.2.3.4/24 gw4 1.2.3.1

Can ruin my internet connection or my servers via Pterodactyl on my server machine?

**Comment by Bazimek** [comment]
I tried it, but it does not work. Still whilst offline error.

**Comment by garrett** [comment]
@ItzMiracleOwO: BTW, I've opened up this more generic issue to better track this problem.

Others have tried the `nmcli` work-around with no success, but have more complicated networks. So it seems to work for some people, but not everyone.

**Comment by garrett** [comment]
@barry-luijten, @kinghat: Which version of Cockpit are you using? If you're not using backports to get Cockpit, it's a version that's 2 years old (version 215). It is my understanding that were some fixes in a more recent version which may be needed for the nmcli workaround to work.

Information on installing from backports:
- on Ubuntu: https://cockpit-project.org/running.html#ubuntu
- on Debian: https://cockpit-project.org/running.html#debian

(The proper fix, of course, is for Debian and Ubuntu to fix their PackageKit integration with NetworkManager and systemd-networkd. I think Ubuntu inherits this bug from Debian, but either could fix the integration in their own distro. Unfortunately, there's not much we can do about it being broken. Hopefully we can find a workaround meanwhile?)

**Comment by kinghat** [comment]
> @barry-luijten, @kinghat: Which version of Cockpit are you using? If you're not using backports to get Cockpit, it's a version that's 2 years old (version 215). It is my understanding that were some fixes in a more recent version which may be needed for the nmcli workaround to work.
> 
> Information on installing from backports:
> 
>     * on Ubuntu: https://cockpit-project.org/running.html#ubuntu
> 
>     * on Debian: https://cockpit-project.org/running.html#debian
> 
> 
> (The proper fix, of course, is for Debian and Ubuntu to fix their PackageKit integration with NetworkManager and systemd-networkd. I think Ubuntu inherits this bug from Debian, but either could fix the integration in their own distro. Unfortunately, there's not much we can do about it being broken. Hopefully we can find a workaround meanwhile?)

```
Cockpit is an interactive Linux server admin interface.
[Project website](https://cockpit-project.org/)

cockpit
    264-1~bpo20.04.1
cockpit-bridge
    264-1~bpo20.04.1
cockpit-machines
    260-1~bpo20.04.1
cockpit-networkmanager
    264-1~bpo20.04.1
cockpit-packagekit
    264-1~bpo20.04.1
cockpit-pcp
    215-1
cockpit-storaged
    264-1~bpo20.04.1
cockpit-system
    264-1~bpo20.04.1
cockpit-ws
    264-1~bpo20.04.1
...

**Comment by garrett** [comment]
BTW, here's the bug reported at PackageKit: https://github.com/PackageKit/PackageKit/issues/336.

It looks like there's a PR that might fix the issue @ https://github.com/PackageKit/PackageKit/pull/506.

It's in PackageKit 1.2.5, which was released 13 days ago. Of course, it's not so likely this will go into an existing LTS distribution release like Debian Stable or Ubuntu LTS. But it's possible that the Stable/LTS versions _might_ cherry-pick this patch.

- Debian Stable has [PackageKit 1.2.2](https://debian.pkgs.org/11/debian-main-amd64/packagekit_1.2.2-2_amd64.deb.html)
- Ubuntu 20.04 LTS has [PackageKit 1.1.13](https://ubuntu.pkgs.org/20.04/ubuntu-updates-main-arm64/packagekit_1.1.13-2ubuntu1.1_arm64.deb.html)

However, it's marked as a critical bugfix @ https://bugs.launchpad.net/ubuntu/+source/packagekit/+bug/1961837

The supposed fix (I haven't tested it) is @ https://github.com/PackageKit/PackageKit/commit/f484808a1d0d14ef3aabcc7eb0a001f3e7830e7e.

It appears to add the following line to `packagekit.service`:

```
Wants=network-online.target
```

**Comment by kinghat** [comment]
> However, it's marked as a critical bugfix @ https://bugs.launchpad.net/ubuntu/+source/packagekit/+bug/1961837
i might be confused but it looks like its fixed in Jammy, the next ubuntu lts release 22.04 :thinking:

**Comment by martinpitt** [comment]
I just [commented](https://github.com/PackageKit/PackageKit/pull/506#pullrequestreview-898527864) on that packagekit change. I don't see how this would work -- this is either a hacky workaround, or (more likely) a no-op. If this works, I'd be very interested to get some confirmation about that.

**Comment by garrett** [comment]
> i might be confused but it looks like its fixed in Jammy, the next ubuntu lts release 22.04 thinking

Yeah, I meant for current Stable / LTS releases, not future ones.

>  I don't see how this would work -- this is either a hacky workaround, or (more likely) a no-op. If this works, I'd be very interested to get some confirmation about that.

Yes, it seems too simple to me too. I guess we'll see:? Hopefully it works (either with this patch or for other reasons). :four_leaf_clover:

**Comment by divStar** [comment]
So... Me experiencing the problem with Ubuntu 22.04LTS and PackageKit 1.2.5 means it's not fixed yet? Though as mentioned in the my bug report calling `pkcon ...` directly seems to always return the desired outcome (though admittedly I only tried a few calls so far).

**Comment by martinpitt** [comment]
@divStar : How exactly do you call `pkcon`? Can you try `pkcon refresh force`? That's using exactly the same API as Cockpit, so it would be a big surprise (and thus well worth looking into) if `refresh force` works, but the cockpit page fails. Thanks!

**Comment by divStar** [comment]
I was calling it from the CLI (ssh into the server).
Sadly it didn't work, apparently for the same reason as Cockpit. But `pkcon get-updates` worked.

Here's the CLI output:
```bash
strange@kamar-taj:~$ sudo pkcon refresh force
[sudo] password for strange:
Refreshing cache              [=========================]
Loading cache                 [=========================]
Finished                      [=========================]
Fatal error: Cannot refresh cache whilst offline
```

**Comment by martinpitt** [comment]
Right, `get-updates` only looks at the current local cache, that works fine. The "am I online" check is only done when PK needs to actualy download something. This is an impedance mismatch between PackageKit asking NetworkManager about online status, and NetworkManager not being responsible for the primary interface (as that is managed by netplan/networkd), and NM not recognizing the latter. IMHO it's a NM bug -- it should recognize that there is a default route, and be satisfied with that. The point is not that it's NM which manages that default route..

**Comment by lfom** [comment]
I have tried all the solutions from #8477, none worked. But the last commend pointed to the correct solution. I install Cockpit on a VPS that uses `cloud-init`, so all network interfaces are listed as "unmanaged" in `nmcli`, so I created a file:
```
$ cat /etc/NetworkManager/conf.d/10-globally-managed-devices.conf                                                                       
[keyfile]
unmanaged-devices=none
```
restarted NetworkManager and it worked fine. I checked all other files, and they were back to the initial config, and I also removed the "fake" network entry as well.

**Comment by RobertoMaurizzi** [comment]
@lfom solution worked for me too: as others, it was a server with a bridged eth0 -> br0 so that the VMs can have real network addresses. Thanks! 👍

**Comment by Ludo-code** [comment]
I have the same problem... I have make one update from cockpit webinterface, and after this i have the same error and all workaround described her doesn't work...

**Comment by garrett** [comment]
@Ludo-code: You're going to have to provide some information for anyone to help you out. Which versions of everything are you using? (Which Linux distribution? Which version of it? Which versions of Cockpit packages? What did you try doing? Etc.)

We have a FAQ about this issue on Ubuntu @ https://cockpit-project.org/faq.html#error-message-about-being-offline

Make sure you're running a currently in support version of your OS. If you're using Ubuntu or Debian, it's probably a good idea to try version in backports instead, which is much more up to date than the default version Ubuntu and Debian ship.

Information about installing Cockpit using backports:
- Ubuntu: https://cockpit-project.org/running.html#ubuntu
- Debian: https://cockpit-project.org/running.html#debian

**Comment by Ludo-code** [comment]
Following our FAQ have solved the problem thanks

**Comment by Dale-CDL** [comment]
I am using the systemd network manager. Do you know if the suggested workaround using a fake network manager still works for me?

**Comment by DmitryYalchik** [comment]
Ubuntu 22.04.4 LTS 
Used `sudo apt install -t jammy-backports cockpit`

First time updating packages from Cockpit was okay but later i begun to getting "Cannot download packages whilst offline" errors

**Comment by frie** [comment]
It seems that the problem finally was solved on the PackageKit side: see [here](https://fedoraproject.org/wiki/PackageKit_Frequently_Asked_Questions#PackageKit_won't_work_without_NetworkManager).
Setting `UseNetworkManager=false` in `/etc/PackageKit/PackageKit.conf` solved the problem on my machines (here: a couple of Debian-based LXC containers)

**Comment by alexrhyshurn** [comment]
Hello,

This solution is still valid in 2026 with debian trixie as both vm and lxc and also with Ubuntu 25.04, however it needs an update.

The IP address 1.2.3.4 is in fact a publicly routable ip address in Australia, I am sure the author gave it as an example, however this is being used all over the web and by machines in production, as a result there is a risk of somebody not being able to reach parts of the public internet.

**TLDR**

Use below command which includes an IP address that is safe:

```
nmcli con add type dummy con-name fake ifname fake0 ip4 192.0.2.2/24 gw4 192.0.2.1
```

**DETAIL:**
The full story is that there are a number of Special Use subnets that can help us:

Dont use the RFC 1918 ones (10.0.0.0, 192.168.0.0 172.16.0.0) as you may be using these internally, but anything from the Documentation/Examples or Testing ranges will be fine.


RFC Number | IP Range | Purpose | CIDR Notation

- RFC 5735 (Obsoleted by RFC 6890) | 10.0.0.0 | Private network addresses | 10.0.0.0/8
- RFC 6890 | 172.16.0.0 | Private network addresses | 172.16.0.0/12
- RFC 6890 | 192.168.0.0 | Private network addresses | 192.168.0.0/16
- RFC 5737 | 192.0.2.0 | Documentation and example addresses | 192.0.2.0/24
- RFC 5737 | 198.51.100.0 | Documentation and example addresses | 198.51.100.0/24
- RFC 5737 | 203.0.113.0 | Documentation and example addresses | 203.0.113.0/24
- RFC 2544 | 198.18.0.0 | Network interconnect device benchmarking | 198.18.0.0/15
- RFC 6890 | 169.254.0.0 | Link-local addresses | 169.254.0.0/16
- RFC 3927 | 127.0.0.0 | Loopback addresses | 127.0.0.0/8
- 

I hope that helps.

Alex

> The Updates page has an "emergent bug" between PackageKit, NetworkManager, and systemd-networkd.
> 
> As seen on
> 
>     * [Ubuntu: packagekit cannot refresh cache whilst offline #8477](https://github.com/cockpit-project/cockpit/issues/8477)
> 
>     * ["Cannot refresh cache whilst offline" error on raspberry pi 4 ( running raspberry pi os) #16333 (comment)](https://github.com/cockpit-project/cockpit/issues/16333#issuecomment-1004070722)
> 
>     * and elsewhere (including other bugs)
> 
> 
> As our users will hit this bug on Ubutntu and Debian systems, we should provide a simple workaround.
> 
> This is not a bug in Cockpit, but this is a UX bug that affects Cockpit.
> 
> The solution is to create a fake network to make all of the subsystems happy:
> 
> ```
> nmcli con add type dummy con-name fake ifname fake0 ip4 1.2.3.4/24 gw4 1.2.3.1
> ```
> 
> We should add the workaround as a simple solution with a short description of the problem and the solution on the Software Updates page when this problem occurs.
> 
> If upstream fixes their systems to not trigger this issue, then it'll be transparently fixed, even with our workaround. Until then, a workaround like this will save a lot of people headaches.

**Comment by EvelynOctavia** [comment]
> The IP address 1.2.3.4 is in fact a publicly routable ip address in Australia

Yes, it's indeed an IP belonging to APNIC Debogon [Project](https://conference.apnic.net/29/pdf/dranse-1slash8.pdf), but the `dummy` interface acts like a loopback for use cases like testing firewall rules, so any packets destined for IP addresses in dummy interface goes back to itself. [Red Hat's documentation on types of network interfaces for virtual networks](https://developers.redhat.com/blog/2018/10/22/introduction-to-linux-interfaces-for-virtual-networking#dummy_interface) uses `1.1.1.1`, which belongs to APNIC for use with Cloudflare DNS :)

Also any update on adding the explanation for doing the workaround? I see you guys discussing on adding an explainer for when `cockpit-ostree` gets installed on a non-`ostree` system. I somewhat waiting for this.

**Comment by martinpitt** [comment]
Thanks @alexrhyshurn ! I sent https://github.com/cockpit-project/cockpit-project.github.io/pull/860 to update the FAQ.

Wrt cockpit-ostree: This FAQ is actually obsolete, the packages now dynamically detect whether they are applicable to the current system ("manifest conditions"). The PR above drops it.

**Comment by Kf637** [comment]
Was a fix ever found for this issue? I've been experiencing it on five of my Ubuntu VMs for quite some time.

**Comment by EvelynOctavia** [comment]
My workaround (and since this gets necro bumped:) Because i only want to 
be able to update my packages and i never touch the network settings on 
my server, what i did is i install cockpit like usual, but with 
`--no-install-recommends`. so the full command is `sudo apt install 
cockpit --no-install-recommends`. that installs cockpit only with it's 
barebones, so it didn't install cockpit-networkmanager. then afterwards 
i install the packagekit component using `sudo apt install 
cockpit-packagekit`. there, i'm able to install updates while missing 
the network tab inside cockpit doesn't matter to me :)
